src/indev/lv_indev.c

```diff
@@ -657,12 +657,33 @@ static void indev_pointer_proc(lv_indev_t * i, lv_indev_data_t * data)
         LV_LOG_WARN("Y is %d which is greater than ver. res", (int)data->point.y);
     }
 
+#if LV_USE_HOVER
+    if((i->pointer.last_point.x != data->point.x || i->pointer.last_point.y != data->point.y)) {
+        /*Move the cursor if set and moved*/
+        if(i->cursor != NULL) lv_obj_set_pos(i->cursor, data->point.x, data->point.y);
+        /*Hover detection*/
+        static lv_obj_t * hv_last[3] = {[0] = NULL, [1] = NULL, [2] = NULL};
+        lv_obj_t * layer[3];   //except layer bottom
+        layer[0] = lv_layer_sys();
+        layer[1] = lv_layer_top();
+        layer[2] = lv_screen_active();
+        lv_obj_t ** last = hv_last, ** l = layer;
+        for(uint8_t h = 0; h < 3; h++, last++) {
+            lv_obj_t * hv = lv_indev_search_obj(*l++, &data->point);
+            if(*last != hv) {
+                if(lv_obj_is_valid(hv)) lv_obj_send_event(hv, LV_EVENT_HOVERED, i);
+                if(lv_obj_is_valid(*last)) lv_obj_send_event(*last, LV_EVENT_HOVER_LEAVE, i);
+                *last = hv;
+            }
+        }
+    }
+#else
```

Comment:
```suggestion
    if((i->pointer.last_point.x != data->point.x || i->pointer.last_point.y != data->point.y)) {
        /*Move the cursor if set and moved*/
        if(i->cursor != NULL) lv_obj_set_pos(i->cursor, data->point.x, data->point.y);
#if LV_USE_HOVER
        /*Hover detection*/
        static lv_obj_t * hv_last[3] = { NULL };
        lv_obj_t * layer[3];   //except layer bottom
        layer[0] = lv_layer_sys();
        layer[1] = lv_layer_top();
        layer[2] = lv_screen_active();
        lv_obj_t * last[3] = hv_last, *l[3] = layer;
        for(uint8_t h = 0; h < 3; h++, last++) {
            lv_obj_t * hv = lv_indev_search_obj(l[h] /* Double check if this's you want */, &data->point);
            if(*last != hv) {
                if(lv_obj_is_valid(hv)) lv_obj_send_event(hv, LV_EVENT_HOVERED, i);
                if(lv_obj_is_valid(*last)) lv_obj_send_event(*last, LV_EVENT_HOVER_LEAVE, i);
                *last = hv;
            }
        }
#endif
    }
```
Remove below `#else` section.

For lvgl v9, we cannot use static variable. You can add them lv_global.

Storing obj's pointer to an array needs to deal with obj delete event to avoid wild pointer.

Inside the for loop, do you mean checking  `if(last[h] != hv) {` instead ?

Reply:
#else section is use because HOVER is checked every movements that position changed of pointer, it may have performance lost even HOVER is not enabled, pointer device will check `if((i->pointer.last_point.x` != data->point.x || i->pointer.last_point.y != data->point.y)) ` if cursor is not set. I think it's reasonable when people who don't need HOVER support within a touch panel only. 
Using `lv_obj_t ** last = hv_last, ** l = layer;` is  to replacing `last[n] ` code, not for any other reasons.
Will try to put `static lv_obj_t * hv_last[3] = { NULL };` them in lv_global as you mentioned. :)